### ISSUE TITLE
Library Panels continued

### ISSUE THREAD

**Issue by hugohaggmark** [grafana/grafana#31305]
Continuation of Epic issue for the Library Panels feature. #29610

It does not show tasks in priority order :)
Please use avatar ![avatar](https://github.com/hugohaggmark.png?size=24) to show which task you're working on:

**Backend**
- [x] Sync Panel Title and Name (#31311) 
- [x] Version column to support optimistic concurrency (#31590)
- [x] Permissions (#31572)
- [x] Improve Library Panel get all API
- [x] Add Type and Description to LibararyPanel table for better filtering possibilities and UX (#32258) 

**Frontend**
- [x] ~~Templating (pause until v8 Panel Edit UX is more polished)~~
   - [x] ~~Create a utility that can identify which variables are used in a library panel (#32206)~~
   - [x] ~~Create a generic utility that can import variables from a dashboard (should be able to reuse this for ordinary Dashboard page as well) (#32206)~~
   - [x] ~~Import Variables dialog (2 steps, chose dashboard first, then chose which variables) (#32206)~~
- [x] Edit Libray Panel in v8 Panel Edit UX branch 
   - [x] Add back search in AddPanelWidget
   - [ ] ~~Add filtering component for Types that can be reused between AddPanelWidget and PanelEdit~~ Moved to issue #32975
- [x] Improve the delete library panel flow with confirmation Modal (#32277)
- [x] Improve Library Panel get all API 
- [x] Add Library Panels to the global Dashboard section (#32876)
- [x] Add back library panel section to side pane (#32923)
- [x] Show Modal warning when changing library panels (#32892)
- [x] Add `Current Panel to Library Panel` to Share Modal as a Tab for normal Panels (#32953)
- [x] Add `Current Panel to Library Panel` to PanelMenu -> More for normal Panels  (#32908)
- [x] Add `Reset to normal Panel` to PanelMenu -> More for Library Panels (#32908)

- [ ] Make switching library panel more robust, currently, the plugin & model is changed in two steps which cause flickering and sometimes crashes as the model is in an inconsistent state . 

**Metrics**
- [ ] ~~Collect usage metrics for Cloud customers~~ Moved to issue #32976
   - How many panels on average are users creating
   - <identify more metrics>

**On hold**
- [ ] Tags (On hold)
- [ ] Starring (Do not think we will ever need this)
- [ ] General Search (on Hold)

**Bugs/Questions**
- [x] Save Panel directly after adding a panel to panel library, shouldn't show all dashboards
![image](https://user-images.githubusercontent.com/562238/108313034-e7fcf280-71b7-11eb-86af-873654fc7761.png)
- [x] Changing visualization removes libarary panel options for a panel
![bug](https://user-images.githubusercontent.com/562238/108313869-47a7cd80-71b9-11eb-8fff-486fef8d5689.gif)
- [ ] Solution for discarding changes for Library Panels when user navigates back, ESC etc
- [x] Save Panel that's only used on the same dashboard, shouldn't show update all instances modal? (#31606)
![image](https://user-images.githubusercontent.com/562238/108313302-66599480-71b8-11eb-8254-fb32c0c16e61.png)
- [x] Should the Library Panel being edited show up in the list of available Library Panels? (https://github.com/grafana/grafana/pull/31472)
![image](https://user-images.githubusercontent.com/562238/108314364-0fed5580-71ba-11eb-9c89-3d9c5d012ef6.png)
- [x] Clicking on `Use instead of current panel` button shiftes the panel plugin but doesn't switch the actual visualization. The video shows switching from Table to Bar Gauge and only after saving the dashboard and refreshing the whole page does the Bar Gauge show up. (https://github.com/grafana/grafana/pull/31643)
![bug](https://user-images.githubusercontent.com/562238/108315176-58594300-71bb-11eb-811a-57c4a8e4d619.gif)

Root issue #1803

**Comment by jessover9000** [comment]
Hey team! Reviewed the status quo of the feature. Some nitpicks (not sure whether just issues with my Grafana or yet to be fixed):

- The terminology is a bit messy, some places say library panel, some places global panel. The user research resulted in "Shared panels" so we should remember to check all the views for the right wording 🤓 Could also change "Add current panel to library" button to "Share current panel to library".
- For me once a panel is added to the library, the panel edit still looks almost identical. The save button text changing is the only feedback I get as a user that I'm not editing a regular panel. In the original design there was supposed to be the reusable panel icon depicted in the top left with the panel name, and also the description "Shared panel" under the title. I recommend that we add that in there
- The "Global panel information" somehow feels out of place within the options dropdown, since those are not options but just information. Could we just show that info either above the dropdowns in the panel settings, or maybe even as part of the subtitle on the left? See a mockup here:
![image](https://user-images.githubusercontent.com/53754871/115887507-49f50500-a452-11eb-811b-e065f5422e77.png)

- Since the shared panels are now part of the visualization picker, I think if you search for something while the picker is expanded search results should include both visualizations and global panels (in the panel settings the search also spans all tabs).
- The visualization picker tooltip should say "Click to change visualization or replace with a shared panel"
- The search box placeholder text should say "Search for visualizations or shared panels"
- The Shared panels / Global panels tab should get an info icon with a tooltip that explains more about what those panels are and links to the docs
- The modal for "Change to library panel", I think we should write "Replace with library panel" because that explains more clearly what happens there.

**Managing the panel library**
- Could we just add a sort dropdown and a filter dropdown under the search box similarly to the manage dashboards view? 
![image](https://user-images.githubusercontent.com/53754871/115887682-76108600-a452-11eb-87fb-838dea1853e9.png)
We might not need the folder vs list view for now, and no blue buttons in the top either, so just those dropdowns left and right. The filter dropdown could be "Filter by visualization type" instead of by tag. That would keep the views sort of consistent for now until we have redesigned the top navigation and adapted the filtering to maybe include tags and other advanced features
- If we introduce a filter there, maybe in the panel edit panel library tab such a filter could also be useful?
- Should panel description be shown in the global panels list? I'm thinking it could be useful to understand better what each panel is about

**Comment by hugohaggmark** [comment]
> Hey team! Reviewed the status quo of the feature. Some nitpicks (not sure whether just issues with my Grafana or yet to be fixed):
> 
> * The terminology is a bit messy, some places say library panel, some places global panel. The user research resulted in "Shared panels" so we should remember to check all the views for the right wording 🤓 Could also change "Add current panel to library" button to "Share current panel to library".
> * For me once a panel is added to the library, the panel edit still looks almost identical. The save button text changing is the only feedback I get as a user that I'm not editing a regular panel. In the original design there was supposed to be the reusable panel icon depicted in the top left with the panel name, and also the description "Shared panel" under the title. I recommend that we add that in there
> * The "Global panel information" somehow feels out of place within the options dropdown, since those are not options but just information. Could we just show that info either above the dropdowns in the panel settings, or maybe even as part of the subtitle on the left? See a mockup here:
>   ![image](https://user-images.githubusercontent.com/53754871/115887507-49f50500-a452-11eb-811b-e065f5422e77.png)
> * Since the shared panels are now part of the visualization picker, I think if you search for something while the picker is expanded search results should include both visualizations and global panels (in the panel settings the search also spans all tabs).
> * The visualization picker tooltip should say "Click to change visualization or replace with a shared panel"
> * The search box placeholder text should say "Search for visualizations or shared panels"
> * The Shared panels / Global panels tab should get an info icon with a tooltip that explains more about what those panels are and links to the docs
> * The modal for "Change to library panel", I think we should write "Replace with library panel" because that explains more clearly what happens there.
> 
> **Managing the panel library**
> 
> * Could we just add a sort dropdown and a filter dropdown under the search box similarly to the manage dashboards view?
>   ![image](https://user-images.githubusercontent.com/53754871/115887682-76108600-a452-11eb-87fb-838dea1853e9.png)
>   We might not need the folder vs list view for now, and no blue buttons in the top either, so just those dropdowns left and right. The filter dropdown could be "Filter by visualization type" instead of by tag. That would keep the views sort of consistent for now until we have redesigned the top navigation and adapted the filtering to maybe include tags and other advanced features
> * If we introduce a filter there, maybe in the panel edit panel library tab such a filter could also be useful?
> * Should panel description be shown in the global panels list? I'm thinking it could be useful to understand better what each panel is about

Thank you @jessover9000, do you have the time to create a new issue out of this?
